client/src/nonblocking/tpu_client.rs

```diff
@@ -0,0 +1,470 @@
+    exit: Arc<AtomicBool>,
+    rpc_client: Arc<RpcClient>,
+}
+
+impl TpuClient {
+    /// Serialize and send transaction to the current and upcoming leader TPUs according to fanout
+    /// size
+    pub async fn send_transaction(&self, transaction: &Transaction) -> bool {
+        let wire_transaction = serialize(transaction).expect("serialization should succeed");
+        self.send_wire_transaction(wire_transaction).await
+    }
+
+    /// Send a wire transaction to the current and upcoming leader TPUs according to fanout size
+    pub async fn send_wire_transaction(&self, wire_transaction: Vec<u8>) -> bool {
+        self.try_send_wire_transaction(wire_transaction)
+            .await
+            .is_ok()
+    }
+
+    /// Serialize and send transaction to the current and upcoming leader TPUs according to fanout
+    /// size
+    /// Returns the last error if all sends fail
+    pub async fn try_send_transaction(&self, transaction: &Transaction) -> TransportResult<()> {
+        let wire_transaction = serialize(transaction).expect("serialization should succeed");
+        self.try_send_wire_transaction(wire_transaction).await
+    }
+
+    /// Send a wire transaction to the current and upcoming leader TPUs according to fanout size
+    /// Returns the last error if all sends fail
+    async fn try_send_wire_transaction(&self, wire_transaction: Vec<u8>) -> TransportResult<()> {
+        let mut last_error: Option<TransportError> = None;
+        let mut some_success = false;
+        for tpu_address in self
+            .leader_tpu_service
+            .leader_tpu_sockets(self.fanout_slots)
+        {
+            // Fake async
+            let result = send_wire_transaction_async(wire_transaction.clone(), &tpu_address);
+            if let Err(err) = result {
+                last_error = Some(err);
+            } else {
+                some_success = true;
+            }
+        }
+        if !some_success {
+            Err(if let Some(err) = last_error {
+                err
+            } else {
+                std::io::Error::new(std::io::ErrorKind::Other, "No sends attempted").into()
+            })
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Create a new client that disconnects when dropped
+    pub async fn new(
+        rpc_client: Arc<RpcClient>,
+        websocket_url: &str,
+        config: TpuClientConfig,
+    ) -> Result<Self> {
+        let exit = Arc::new(AtomicBool::new(false));
+        let leader_tpu_service =
+            LeaderTpuService::new(rpc_client.clone(), websocket_url, exit.clone()).await?;
+
+        Ok(Self {
+            fanout_slots: config.fanout_slots.min(MAX_FANOUT_SLOTS).max(1),
+            leader_tpu_service,
+            exit,
+            rpc_client,
+        })
+    }
+
+    pub async fn send_and_confirm_messages_with_spinner<T: Signers>(
+        &self,
+        messages: &[Message],
+        signers: &T,
+    ) -> Result<Vec<Option<TransactionError>>> {
```

Comment:
This one's all copied, no changes except for `await` wherever needed.  Any ideas for how to improve it are welcome

Reply:
> exposing the underlying tokio runtime and blocking on async operations, but send_transactions_with_spinner would hog the runtime for too long to make sense

Why would `send_transactions_with_spinner()` hog the runtime here, I'd think we'd always used the multithreaded runtime right?

Reply:
At a 1million foot level, I'm pretty satisfied with how the blocking RpcClient was written in terms of the non-blocking RpcClient. Is that simply a non-starter for TpuClient because some reason?

Reply:
Definitely agreed that it's nice to have the blocking client use the nonblocking version, so evidently I haven't explained the reason well :sweat: 

Essentially, it's possible, but it would break TPU client users in multi-threaded contexts.

Reply:
i think it could be argued that the `*_with_spinner` variants only belong in the sync API since they're presumably operating on human timescales? seems like with that we could better handle presentation of multiple pending transactions than we do today, which is to just let a bunch of threads stomp all over each other's stdout

Reply:
There's still an important use-case for  this-- if someone's in async-land and wants to send a bunch of transactions, it's important to have some form of `send_messages_with_retry`.  I do agree on the std-out mess... would it be better to implement a version of `*_with_spinner` that takes in some sink that it can write log messages to rather than using the progress bar directly?